### PR TITLE
[SW-1493] Properly apply type checks even for None values on all H2OAutoML parameters

### PR DESCRIPTION
**Pull request by jakubhava**

### CODE REVIEW

File: py/ai/h2o/sparkling/ml/algos/H2OAutoML.py [hunk 1]
```diff
@@ -16,52 +16,26 @@
...
+from ai.h2o.sparkling.ml.algos.H2OAlgoBase import H2OAlgoBase
 from ai.h2o.sparkling.ml.params import H2OAutoMLParams
-from ai.h2o.sparkling.sparkSpecifics import get_input_kwargs
 
 
-class H2OAutoML(H2OAutoMLParams, H2OAutoMLExtensions, JavaEstimator, JavaMLReadable, JavaMLWritable):
+class H2OAutoML(H2OAutoMLParams, H2OAlgoBase):
 
     @keyword_only
-    def __init__(self, featuresCols=[], labelCol="label", allStringColumnsToCategorical=True, columnsToCategorical=[], splitRatio=1.0, foldCol=None,
```
Comment (jakubhava) [inline py/ai/h2o/sparkling/ml/algos/H2OAutoML.py]: We can actually optimize the code like this so we don't duplicate any information from Java side

File: py/ai/h2o/sparkling/ml/algos/H2OAutoML.py [hunk 2]
```diff
@@ -16,52 +16,26 @@
...
-        kwargs = get_input_kwargs(self)
-
-        return self._set(**kwargs)
-
-    def _create_model(self, java_model):
-        return H2OMOJOModel(java_model)
+        self._setDefaultValuesFromJava()
+        self._set(**kwargs)
+
+    def leaderboard(self):
```
Comment (jakubhava) [inline py/ai/h2o/sparkling/ml/algos/H2OAutoML.py]: Moved back as I don't think there is special need to generate the Algos now as the code is really short

File: py/ai/h2o/sparkling/ml/algos/H2OAutoML.py [hunk 3]
```diff
@@ -16,52 +16,26 @@
...
 
     @keyword_only
-    def __init__(self, featuresCols=[], labelCol="label", allStringColumnsToCategorical=True, columnsToCategorical=[], splitRatio=1.0, foldCol=None,
-                 weightCol=None, ignoredCols=[], includeAlgos=None, excludeAlgos=None, projectName=None, maxRuntimeSecs=3600.0, stoppingRounds=3,
-                 stoppingTolerance=0.001, stoppingMetric="AUTO", nfolds=5, convertUnknownCategoricalLevelsToNa=True, seed=-1,
-                 sortMetric="AUTO", balanceClasses=False, classSamplingFactors=None, maxAfterBalanceSize=5.0,
-                 keepCrossValidationPredictions=True, keepCrossValidationModels=True, maxModels=0,
-                 predictionCol="prediction", detailedPredictionCol="detailed_prediction", withDetailedPredictionCol=False,
-                 convertInvalidNumbersToNa=False, **deprecatedArgs):
+    def __init__(self, **kwargs):
```
Comment (mn-mikke) [inline py/ai/h2o/sparkling/ml/algos/H2OAutoML.py]: The same concern as [here](https://github.com/h2oai/sparkling-water/pull/1426/files#r312550016).

File: py/ai/h2o/sparkling/ml/params/H2OAutoMLParams.py [hunk 4]
```diff
@@ -27,26 +26,26 @@ class H2OAutoMLParams(H2OCommonSupervisedParams):
...
                        H2OTypeConverters.toEnumString("ai.h2o.sparkling.ml.algos.H2OAutoMLSortMetric"))
-    balanceClasses = Param(Params._dummy(), "balanceClasses", "Balance classes", TypeConverters.toBoolean)
-    classSamplingFactors = Param(Params._dummy(), "classSamplingFactors", "Class sampling factors", TypeConverters.toListFloat)
-    maxAfterBalanceSize = Param(Params._dummy(), "maxAfterBalanceSize", "Max after balance size", TypeConverters.toFloat)
-    keepCrossValidationPredictions = Param(Params._dummy(), "keepCrossValidationPredictions", "Keep cross validation predictions", TypeConverters.toBoolean)
-    keepCrossValidationModels = Param(Params._dummy(), "keepCrossValidationModels", "Keep cross validation models", TypeConverters.toBoolean)
-    maxModels = Param(Params._dummy(), "maxModels", "Max models to train in AutoML", TypeConverters.toInt)
+    balanceClasses = Param(Params._dummy(), "balanceClasses", "Balance classes", H2OTypeConverters.toBoolean())
+    classSamplingFactors = Param(Params._dummy(), "classSamplingFactors", "Class sampling factors", H2OTypeConverters.toNullableListFloat())
```
Comment (mn-mikke) [inline py/ai/h2o/sparkling/ml/params/H2OAutoMLParams.py]: Please can help me understand why `classSamplingFactors` can contains `null`s?

File: py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py [hunk 5]
```diff
@@ -25,16 +26,177 @@ class H2OTypeConverters(object):
...
+
+        return convert
+
+    @staticmethod
+    def toNullableString():
+        def convert(value):
+            if value is None:
+                return None
+            else:
+                return H2OTypeConverters.toString()(value)
```
Comment (mn-mikke) [inline py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py]: WDYT about calling `TypeConverters.toString(value)` directly?

File: py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py [hunk 6]
```diff
@@ -25,16 +26,177 @@ class H2OTypeConverters(object):
...
+
+        return convert
+
+    @staticmethod
+    def toNullableInt():
+        def convert(value):
+            if value is None:
+                return None
+            else:
+                return H2OTypeConverters.toInt()(value)
```
Comment (mn-mikke) [inline py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py]: ditto

File: py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py [hunk 7]
```diff
@@ -25,16 +26,177 @@ class H2OTypeConverters(object):
...
+
+        return convert
+
+    @staticmethod
+    def toNullableFloat():
+        def convert(value):
+            if value is None:
+                return None
+            else:
+                return H2OTypeConverters.toFloat()(value)
```
Comment (mn-mikke) [inline py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py]: ditto

File: py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py [hunk 8]
```diff
@@ -25,16 +26,177 @@ class H2OTypeConverters(object):
...
+
+        return convert
+
+    @staticmethod
+    def toNullableBoolean():
+        def convert(value):
+            if value is None:
+                return None
+            else:
+                return H2OTypeConverters.toBoolean()(value)
```
Comment (mn-mikke) [inline py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py]: ditto

File: py/ai/h2o/sparkling/ml/algos/H2OAutoML.py (hunk 3 above)
Comment (jakubhava) [reply py/ai/h2o/sparkling/ml/algos/H2OAutoML.py]: Was thinking about this and still wasn't sure what was better win. 

We can still generate the code, but have the arguments mentioned just once in the ``__init__ `` method. The 2 other just duplicates already specified information

File: py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py (hunk 5 above)
Comment (jakubhava) [reply py/ai/h2o/sparkling/ml/params/H2OTypeConverters.py]: good point, done

File: py/ai/h2o/sparkling/ml/params/H2OAutoMLParams.py (hunk 4 above)
Comment (jakubhava) [reply py/ai/h2o/sparkling/ml/params/H2OAutoMLParams.py]: Class sampling factors can contain `None` as H2O's code distinguished between null and empty array. The elements itself can't be null

File: py/ai/h2o/sparkling/ml/algos/H2OAutoML.py (hunk 3 above)
Comment (jakubhava) [reply py/ai/h2o/sparkling/ml/algos/H2OAutoML.py]: I put the arguments back into init method, we can iterate later, but we can probably generate it later

**Pull request merged**